src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+import React, {
+  createRef,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+  forwardRef,
+} from 'react';
+import PropTypes from 'prop-types';
+import FormControl from '../form-control/form-control.jsx';
+import FormControlIcons from '../form-control-icons/form-control-icons.jsx';
+import Icon from '../icon/icon.jsx';
+import IconButton from '../icon-button/icon-button.jsx';
+import iconCaretDown from '../../svgs/caret-down.svg';
+import iconClear from '../../svgs/clear.svg';
+import Listbox from '../listbox/listbox.jsx';
+import ListOption from '../list-option/list-option.jsx';
+import Loader from '../loader/loader.jsx';
+import NoOptions from '../no-options/no-options.jsx';
+import TagList from '../tag-list/tag-list.jsx';
+import Tag from '../tag/tag.jsx';
+import useDropdownClose from '../../hooks/use-dropdown-close.js';
+import styles from './multi-select.css';
+
+const MultiSelect = forwardRef(function MultiSelect(props, ref) {
+  const [autocompleteValue, setAutocompleteValue] = useState('');
+  const autocompleteRef = useRef();
+  const backupRef = useRef();
+  const [expanded, setExpanded] = useState(false);
+  const selfRef = ref || backupRef;
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(props.value || []);
+  const valueRefs = value.map(() => createRef());
+  const visibleOptions = getVisibleOptions();
+  const visibleOptionsRefs = visibleOptions.map(() => createRef());
+  const wrapperRef = useRef(null);
+
+  const ids = {
+    emptyMessage: `${props.id}-empty`,
+    label: `${props.id}-label`,
+    listbox: `${props.id}-listbox`,
+    textbox: `${props.id}-autocomplete`,
+  };
+
+  const classNames = {
```

Comment:
We can "streamline" this:

```js
const classNames = {
  ...MultiSelect.defaultProps.classNames,
  ...props.classNames
}
```

where

```js
MultiSelect.defaultProps = {
  classNames: {
    container: ....,
    formControlChildrenContainer: ...,
     ...
  {
};
```

Reply:
Note: while we discussed this in 1:1 review, we realized this approach doesn't work, because we need the component context to determine classNames. Further exploration of changes to this pattern to support a better `defaultProps` declaration is needed.